File: testdata/fixtures/elixir/CWE-89.ex
# CWE-89: SQL Injection – user input concatenated into SQL query.
# Ecto query with variable is vulnerable; literals are safe.

defmodule Fixture.CWE89 do
  def vuln(user_input) do
    # ruleid: elixir-sql-injection
    Ecto.Adapters.SQL.query(Repo, "select * from users where id = " <> user_input)
    # ruleid: elixir-sql-injection
    Repo.query("select * from t where name = " <> user_input)
    # ruleid: elixir-sql-injection
    Ecto.Adapters.SQL.query(Repo, "select * from t where x = " <> user_input)
    # ruleid: elixir-sql-injection
    q = user_input; Repo.query("select * from t where id = " <> q)
  end

  def safe do
    # ok: elixir-sql-injection
    Ecto.Adapters.SQL.query(Repo, "select * from users where id = 1")
    # ok: elixir-sql-injection
    Repo.query("select * from users")
    # ok: elixir-sql-injection
    Repo.query("select * from users where id = $1", [1])
  end
end

File: testdata/fixtures/ocaml/CWE-89.ml
(* CWE-89: SQL Injection – user input in DB query without sanitization.
 * Db.query with variable is vulnerable; literals are safe.
 *)

let vuln user_input =
  (* ruleid: ocamllint-sqli *)
  let _ = Db.query user_input in
  (* ruleid: ocamllint-sqli *)
  let _ = Mysql.query user_input in
  (* ruleid: ocamllint-sqli *)
  let _ = Db.query (user_input ^ " LIMIT 1") in
  (* ruleid: ocamllint-sqli *)
  let q = user_input in let _ = Mysql.query q in
  ()

let safe () =
  (* ok: ocamllint-sqli *)
  let _ = Db.query "SELECT * FROM users WHERE id = 1" in
  (* ok: ocamllint-sqli *)
  let _ = Mysql.query "SELECT 1" in
  (* ok: ocamllint-sqli *)
  let _ = Db.query "SELECT * FROM t WHERE name = ?" in
  ()
